lib/msal-common/src/client/SilentFlowClient.ts

```diff
@@ -0,0 +1,161 @@
+            uniqueId: idTokenObj.claims.oid || idTokenObj.claims.sub,
+            tenantId: idTokenObj.claims.tid,
+            scopes: requestScopes.asArray(),
+            account: CacheHelper.toIAccount(cacheRecord.account),
+            idToken: cacheRecord.idToken.secret,
+            idTokenClaims: idTokenObj.claims,
+            accessToken: cacheRecord.accessToken.secret,
+            expiresOn: new Date(cacheRecord.accessToken.expiresOn),
+            extExpiresOn: new Date(cacheRecord.accessToken.extendedExpiresOn),
+            familyId: null,
+        };
+    }
+
+    /**
+     * fetches idToken from cache if present
+     * @param request
+     */
+    private fetchIdToken(homeAccountId: string, environment: string, inputRealm: string): IdTokenEntity {
+        const idTokenKey: string = CacheHelper.generateCredentialCacheKey(
+            homeAccountId,
+            environment,
+            CredentialType.ID_TOKEN,
+            this.config.authOptions.clientId,
+            inputRealm
+        );
+        return this.unifiedCacheManager.getCredential(idTokenKey) as IdTokenEntity;
+    }
+
+    /**
+     * fetches accessToken from cache if present
+     * @param request
+     * @param scopes
+     */
+    private fetchAccessToken(homeAccountId: string, environment: string, scopes: ScopeSet, inputRealm: string): AccessTokenEntity {
+        const accessTokenFilter: CredentialFilter = {
+            homeAccountId,
+            environment,
+            credentialType: CredentialType.ACCESS_TOKEN,
+            clientId: this.config.authOptions.clientId,
+            realm: inputRealm,
+            target: scopes.printScopes()
+        };
+        const credentialCache: CredentialCache = this.unifiedCacheManager.getCredentialsFilteredBy(accessTokenFilter);
+        const accessTokens = Object.values(credentialCache);
+        if (accessTokens.length > 1) {
+            // TODO: Figure out what to throw or return here.
+        } else if (accessTokens.length < 1) {
+            return null;
+        }
+        return accessTokens[0] as AccessTokenEntity;
+    }
+
+    /**
+     * fetches refreshToken from cache if present
+     * @param request
+     */
+    private fetchRefreshToken(homeAccountId: string, environment: string): RefreshTokenEntity {
+        const refreshTokenKey: string = CacheHelper.generateCredentialCacheKey(
+            homeAccountId,
+            environment,
+            CredentialType.REFRESH_TOKEN,
+            this.config.authOptions.clientId
+        );
+        return this.unifiedCacheManager.getCredential(refreshTokenKey) as RefreshTokenEntity;
+    }
+
+    /**
+     * check if an access token is expired
+     * @param expiresOn
+     */
+    private isTokenExpired(expiresOn: string): boolean {
+        // check for access token expiry
+        const expirationSec = Number(expiresOn);
+        const offsetCurrentTimeSec = TimeUtils.nowSeconds() + this.config.systemOptions.tokenRenewalOffsetSeconds;
```

Comment:
Should this be `TimeUtils.nowSeconds() - this.config.systemOptions.tokenRenewalOffsetSeconds;`? Otherwise you will return expired tokens.

Reply:
The offset is used to renew the tokens at a specified amount of time before the actual expiration. I think the actual boolean check needs to be flipped, but the logic itself is correct.